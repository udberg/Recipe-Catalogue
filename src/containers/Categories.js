import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Category from '../components/Category';
import grid from '../styles/grid.css';

const cx = classNames.bind(grid);

const Categories = ({ categories }) => (
  <div className={cx('grid', 'grid-column', 'grid-gap-1/5')}>
    {categories ? categories.map((category) => <Category key={category.idCategory} category={category} />) : 'No categories'}
  </div>
);

Categories.defaultProps = {
  categories: [],
};

Categories.propTypes = {
  categories: PropTypes.arrayOf(PropTypes.object),
};

const mapStateToProps = (state) => ({
  categories: state.categories,
});

export default connect(mapStateToProps)(Categories);
