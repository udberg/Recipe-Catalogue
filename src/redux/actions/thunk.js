import { fetchMeals, fetchMeal, fetchSearchMeals } from '../../API/api';
import {
  loadingMeal, errorMeal, successMeal, loadingMeals, errorMeals, successMeals,
} from './index';

const fetchMealsByCategory = (category) => async (dispatch) => {
  dispatch(loadingMeals());
  const response = fetchMeals(category);
  try {
    const data = await response;
    return dispatch(successMeals(data.meals));
  } catch (e) {
    return dispatch(errorMeals('Error while fetching data.'));
  }
};

const searchMeals = (query) => async (dispatch) => {
  dispatch(loadingMeals());
  const response = fetchSearchMeals(query);
  try {
    const data = await response;
    if (data.meals) {
      return dispatch(successMeals(data.meals));
    }
    return dispatch(errorMeals(`There is no meal with name: ${query}`));
  } catch (e) {
    return dispatch(errorMeals('Error while fetching data.'));
  }
};

const fetchMealById = (id) => async (dispatch) => {
  dispatch(loadingMeal());
  const response = fetchMeal(id);
  try {
    const data = await response;
    return dispatch(successMeal(data.meals[0]));
  } catch (e) {
    return dispatch(errorMeal('Error while fetching data.'));
  }
};

export { fetchMealsByCategory, searchMeals, fetchMealById };
