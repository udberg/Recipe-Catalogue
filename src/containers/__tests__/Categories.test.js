import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import ConfigureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Categories from '../Categories';

const mockStore = ConfigureMockStore([thunk]);

let store;

beforeEach(() => {
  store = mockStore({});
});

test('should test for empty category', () => {
  const { container } = render(
    <Provider store={store}>
      <MemoryRouter>
        <Categories />
      </MemoryRouter>
    </Provider>,
  );

  expect(container.textContent).toMatch('');
});

test('should match the snapshot', () => {
  store = mockStore({
    categories: [
      {
        idCategory: '1',
        strCategory: 'Beef',
        strCategoryThumb: 'https://www.themealdb.com/images/category/beef.png',
        strCategoryDescription: 'Beef is the culinary name for meat from cattle, particularly skeletal muscle. Humans have been eating beef since prehistoric times.[1] Beef is a source of high-quality protein and essential nutrients.[2]',
      },
      {
        idCategory: '2',
        strCategory: 'Chicken',
        strCategoryThumb: 'https://www.themealdb.com/images/category/chicken.png',
        strCategoryDescription: 'Chicken is a type of domesticated fowl, a subspecies of the red junglefowl. It is one of the most common and widespread domestic animals, with a total population of more than 19 billion as of 2011.[1] Humans commonly keep chickens as a source of food (consuming both their meat and eggs) and, more rarely, as pets.',
      },
    ],
  });
  const { container } = render(
    <Provider store={store}>
      <MemoryRouter>
        <Categories />
      </MemoryRouter>
    </Provider>,
  );

  expect(container).toMatchSnapshot();
});
