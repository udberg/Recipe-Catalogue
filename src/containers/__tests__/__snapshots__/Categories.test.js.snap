// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div>
  <div
    class="grid grid-column grid-gap-1/5"
  >
    <div
      class="card"
    >
      <a
        class="card-link"
        href="/meals/Beef"
      >
        <img
          alt="Beef"
          class="card-img"
          src="https://www.themealdb.com/images/category/beef.png"
        />
        <div
          class="card-body"
        >
          <h3
            class="card-title"
          >
            Beef
          </h3>
        </div>
      </a>
    </div>
    <div
      class="card"
    >
      <a
        class="card-link"
        href="/meals/Chicken"
      >
        <img
          alt="Chicken"
          class="card-img"
          src="https://www.themealdb.com/images/category/chicken.png"
        />
        <div
          class="card-body"
        >
          <h3
            class="card-title"
          >
            Chicken
          </h3>
        </div>
      </a>
    </div>
  </div>
</div>
`;
